libraries/libraries.go

```diff
@@ -0,0 +1,208 @@
+package libraries
+
+import (
+	"github.com/src-d/go-borges"
+	"github.com/src-d/go-borges/util"
+
+	"gopkg.in/src-d/go-errors.v1"
+)
+
+var (
+	// ErrLibraryExists an error returned when a borges.Library
+	// added before is attempted to be added again.
+	ErrLibraryExists = errors.NewKind("library %s already exists")
+)
+
+// FilterLibraryFunc stands for a borges.Library filter function.
+type FilterLibraryFunc func(borges.Library) (bool, error)
+
+// Libraries is an implementation to aggregate borges.Library in just one instance.
+// The borges.Library that will be added shouldn't contain other libraries inside.
+type Libraries struct {
+	libs map[borges.LibraryID]borges.Library
+}
+
+var _ borges.Library = (*Libraries)(nil)
+
+// New create a new Libraries instance.
+func New() *Libraries {
+	return &Libraries{
+		libs: map[borges.LibraryID]borges.Library{},
+	}
+}
+
+// Add adds a new borges.Library. It will fail with ErrLibraryExists
+// if the library was already added.
+func (l *Libraries) Add(lib borges.Library) error {
+	_, ok := l.libs[lib.ID()]
+	if ok {
+		return ErrLibraryExists.New(lib.ID())
+	}
+
+	l.libs[lib.ID()] = lib
+	return nil
+}
+
+// ID implements the Library interface.
+func (l *Libraries) ID() borges.LibraryID {
+	return ""
+}
+
+// Init implements the Library interface.
+func (l *Libraries) Init(borges.RepositoryID) (borges.Repository, error) {
+	return nil, borges.ErrNotImplemented.New()
+}
+
+// Get implements the Library interface.
+func (l *Libraries) Get(id borges.RepositoryID, mode borges.Mode) (borges.Repository, error) {
+	var repos []borges.Repository
+	for _, lib := range l.libs {
+		r, err := lib.Get(id, mode)
+		if err != nil {
+			if borges.ErrRepositoryNotExists.Is(err) {
+				continue
+			}
+
+			return nil, err
+		}
+
+		repos = append(repos, r)
```

Comment:
Why is the repository not returned as soon as it is found?

Reply:
Not sure if we assume that repositories ID are unique among libraries, so it's just to notice that could be more than one found repository. In that case, every time the returned value could be different because libraries are not always iterated in the same order.

If this assumption is wrong, I'll change it to return the first value.

Reply:
But in this code we are not doing anything with the rest of the of the repos. I would assume that repos exists only once in one of the libraries. Besides that, checking all libraries can be very expensive.